PeriodPals/periodpals
# Feat/misc/notifications/simple-interface : Implement push notification simple interface

# Implement push notification simple interface


## Description
This PR introduces complementary functions in `PushNotificationsServiceImpl` that allows us to handle base notification actions. It closes issue #238.

## Changes
This PR essentially overrides the `FirebaseNotificationService` functions to customize the appearance and behaviour of our notifications. 

It also provides a Supabase Edge Function and Webhook that send a notification to a specified user or group of users when a new row is inserted into the "notifications" table in Supabase. For now, this class does not communicate with this Supabase table since we don't yet have a notifications model, but the behaviour is tested and works. A user's device is identified by a "fcm token" that is then stored in the "public.users" table in Supabase.

The notifications in themselves have a title, a description and an icon. The icon is defined in the resources.

**Note:** please note that the TODOs are left in intentionally since they are reserved for future implementations.

## Files 

#### Added
- all the `ic_notification_icon.png`s in the `res` folder: they are all the same icon but different sizes (5 files)
- `supabase/config.toml`: our Supabase configuration when running locally
- `supabase/functions/push/index.ts`: the edge function used to trigger notifications
- `supabase/.gitignore`: associated .gitignore

#### Modified
- `app/src/main/java/com/android/periodpals/MainActivity.kt`: implemented FCM token creation and proper initialization of Google Play Services, on which FCM relies
- `app/src/main/java/com/android/periodpals/services/PushNotificationsServiceImpl.kt`: overrode FCM methods to customize them, implemented the creation of a notification channel when a user grants the notification permissions and the creation of device tokens
- `app/src/test/java/com/android/periodpals/services/PushNotificationsServiceImplTest.kt`: added tests to check the new functionalities (more details in the Testing section)
- `gradle/libs.versions.toml`, `app/build.gradle.kts`: added Google Play Services dependency
- `app/src/main/AndroidManifest.xml`: added explicit permissions to comply with good practices and needs for the Google Play Services

#### Removed
None

## Dependencies Added
- `libs.play.services.base`: Google Play Services dependency

## Testing
As the model for notifications is not implemented and integrated in this file yet, the added tests are quite light. However this should be solved when integrating said functionalities.

To test for yourself the notification from your browser:
 - run the app on your device and accept the notification permissions (if you don't have the permissions request dialog, run Gradle's `uninstallAll` and retry)
 - in the logcat, search for your device token and save it **or** search for your account in the public.users table in Supabase and get the "fcm_token" value
 - open your Firebase console and go to Messaging in the left sidebar
 - click on Create your first campaign then Firebase Notification messages
 - enter your title and text, then before clicking Next, click on Send test message and enter your device token
 - click on Next and select the PeriodPals app
 - Next Next and Next again, since the fields are optional
 - click on Review > Publish

You should now receive a notification and have a new line added to the notifications table in Supabase. Note that you should have the app running in the background to get the notification.
